homeassistant/components/obihai/config_flow.py

```diff
@@ -16,60 +20,135 @@
-        validate_auth,
-        user_input[CONF_HOST],
-        user_input[CONF_USERNAME],
-        user_input[CONF_PASSWORD],
-    )
+
+    if user_input[CONF_USERNAME] and user_input[CONF_PASSWORD]:
+        return await hass.async_add_executor_job(
+            validate_auth,
+            user_input[CONF_HOST],
+            user_input[CONF_USERNAME],
+            user_input[CONF_PASSWORD],
+        )
+
+    # Don't bother authenticating if we've already determined the credentials are invalid
+    return None
 
 
 class ObihaiFlowHandler(ConfigFlow, domain=DOMAIN):
     """Config flow for Obihai."""
 
-    VERSION = 1
+    VERSION = 2
+    discovery_schema: vol.Schema | None = None
+    _dhcp_discovery_info: dhcp.DhcpServiceInfo | None = None
 
     async def async_step_user(
         self, user_input: dict[str, Any] | None = None
     ) -> FlowResult:
         """Handle a flow initialized by the user."""
+
         errors: dict[str, str] = {}
+        ip: str | None = None
 
         if user_input is not None:
-            self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST]})
-            if await async_validate_creds(self.hass, user_input):
-                return self.async_create_entry(
-                    title=user_input[CONF_HOST],
-                    data=user_input,
-                )
-            errors["base"] = "cannot_connect"
+            try:
+                ip = gethostbyname(user_input[CONF_HOST])
+            except gaierror:
+                errors["base"] = "cannot_connect"
 
-        data_schema = self.add_suggested_values_to_schema(DATA_SCHEMA, user_input)
+            if ip:
+                if pyobihai := await async_validate_creds(self.hass, user_input):
+                    device_mac = await self.hass.async_add_executor_job(
+                        pyobihai.get_device_mac
+                    )
+                    await self.async_set_unique_id(device_mac)
+                    self._abort_if_unique_id_configured()
+
+                    return self.async_create_entry(
+                        title=user_input[CONF_HOST],
+                        data=user_input,
+                    )
+                errors["base"] = "invalid_auth"
+
+        data_schema = self.discovery_schema or DATA_SCHEMA
         return self.async_show_form(
             step_id="user",
             errors=errors,
-            data_schema=data_schema,
+            data_schema=self.add_suggested_values_to_schema(data_schema, user_input),
         )
 
+    async def async_step_dhcp(self, discovery_info: dhcp.DhcpServiceInfo) -> FlowResult:
+        """Prepare configuration for a DHCP discovered Obihai."""
+
+        self._dhcp_discovery_info = discovery_info
+        return await self.async_step_dhcp_confirm()
+
+    async def async_step_dhcp_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Attempt to confirm."""
+
+        if self._dhcp_discovery_info:
```

Comment:
Is this check to make mypy happy?

If it is, you can add an `assert self._dhcp_discovery_info` instead since it should be possible to end up here without having `self._dhcp_discovery_info` set.

Reply:
Yes, that was the reason. Updated.